File: src/frontend/src/app/features/recipes/models/recipe.models.ts
export interface RecipeIngredientRequest {
  ingredientId: string;
  name: string;
  quantity: number;
  unit: string;
  note: string | null;
}

export interface CreateRecipeRequest {
  title: string;
  descriptionMarkdown: string | null;
  /** Number of adult portions the recipe yields. One serving is a single adult portion. */
  servings: number;
  mealType?: RecipeMealType | null;
  ingredients: RecipeIngredientRequest[];
  /** Original source of the recipe (e.g. the video or web page it was imported from). */
  sourceUrl?: string | null;
  /** Marks the recipe as created with AI assistance; ignored on update (provenance is kept). */
  isAiAssisted?: boolean;
}

export type RecipeMealType = 'breakfast' | 'second-breakfast' | 'dinner' | 'snack' | 'supper';
export const recipeMealTypes: { value: RecipeMealType; label: string }[] = [
  { value: 'breakfast', label: 'meal.slot.breakfast' },
  { value: 'second-breakfast', label: 'meal.slot.second-breakfast' },
  { value: 'dinner', label: 'meal.slot.dinner' },
  { value: 'snack', label: 'meal.slot.snack' },
  { value: 'supper', label: 'meal.slot.supper' },
];

export type UpdateRecipeRequest = CreateRecipeRequest;

export interface RecipeIngredient {
  id: string;
  sortOrder: number;
  ingredientId: string;
  name: string;
  quantity: number;
  unit: string;
  normalizedQuantity: number;
  normalizedUnit: string;
  note: string | null;
}

export type RecipeAuditImportance = 'important' | 'minor';
export type RecipeAuditElement = 'ingredients' | 'description' | 'image' | 'mealType';

export interface RecipeAuditItem {
  id: string;
  title: string;
  slug: string;
  importance: RecipeAuditImportance;
  missingElements: RecipeAuditElement[];
  missingCount: number;
}

export interface RecipeSummary {
  id: string;
  title: string;
  slug: string;
  /** User id of the household member who created the recipe. */
  createdByUserId: string;
  ingredientCount: number;
  servings: number;
  mealType: RecipeMealType | null;
  firstImageUrl: string | null;
  isAiAssisted: boolean;
  createdAtUtc: string;
  updatedAtUtc: string;
}

export interface RecipeDetail {
  id: string;
  title: string;
  slug: string;
  /** User id of the household member who created the recipe. */
  createdByUserId: string;
  descriptionMarkdown: string | null;
  servings: number;
  mealType: RecipeMealType | null;
  ingredients: RecipeIngredient[];
  images: RecipeImage[];
  canEdit: boolean;
  isAiAssisted: boolean;
  sourceUrl: string | null;
  createdAtUtc: string;
  updatedAtUtc: string;
}

export interface RecipeImage {
  id: string;
  recipeId: string;
  fileName: string;
  contentType: string;
  fileSizeBytes: number;
  altText: string | null;
  sortOrder: number;
  contentUrl: `/api/${string}`;
  createdAtUtc: string;
  /** Attribution for the image's primary source; null when the image has no recorded source. */
  source?: ImageSourceAttribution | null;
}

export interface RecipeImageCandidate {
  provider: string;
  externalImageId: string;
  previewUrl: string;
  sourcePageUrl: string;
  authorName: string | null;
  authorUrl: string | null;
  altText: string | null;
  width: number | null;
  height: number | null;
}

export interface RecipeImageImportResult {
  content: string;
  contentType: 'image/webp';
  width: number;
  height: number;
  provider: string;
  externalImageId: string;
  sourcePageUrl: string;
  authorName: string | null;
  authorUrl: string | null;
  altText: string | null;
}

export interface RecipeImageSourceData {
  provider: string;
  externalId: string | null;
  url: string | null;
  authorName: string | null;
  authorUrl: string | null;
}

export interface ImageSourceAttribution {
  provider: string;
  authorName: string | null;
  authorUrl: string | null;
  url: string | null;
}

export interface PagedResponse<T> {
  items: T[];
  page: number;
  pageSize: number;
  totalCount: number;
}

export enum RecipeImportStatus {
  Pending,
  FetchingTranscript,
  Extracting,
  ResolvingIngredients,
  ReadyForReview,
  Failed,
}

export interface RecipeImportIngredient {
  name: string;
  quantity: number | null;
  unit: string | null;
  note: string | null;
  ingredientId: string | null;
  matchedName: string | null;
  isProposedNew: boolean;
}

export interface RecipeImportDraft {
  title: string;
  servings: number;
  instructionsMarkdown: string;
  gaps: string[];
  ingredients: RecipeImportIngredient[];
  duplicateMatches: RecipeExistenceMatch[];
}

export interface RecipeExistenceMatch {
  recipeId: string;
  title: string;
  sourceUrl: string | null;
  matchType: number;
}

export interface RecipeImportJob {
  id: string;
  status: RecipeImportStatus;
  draft: RecipeImportDraft | null;
  errorReason: string | null;
}

File: src/backend/Kotlet.Application/Recipes/RecipeModels.cs
using Kotlet.Domain.Common;
using Kotlet.Domain.MealPlanner;
using Kotlet.Domain.Recipes;

namespace Kotlet.Application.Recipes;

public sealed record RecipeIngredientRequest(
    Guid IngredientId,
    decimal Quantity,
    string Unit,
    string? Note);

public sealed record CreateRecipeRequest(
    string Title,
    string? DescriptionMarkdown,
    IReadOnlyList<RecipeIngredientRequest> Ingredients,
    int Servings = 1,
    string? MealType = null,
    string? SourceUrl = null,
    bool IsAiAssisted = false,
    string? VideoUrl = null,
    string? VideoThumbnailUrl = null);

// No IsAiAssisted here: the flag records provenance and survives human edits.
public sealed record UpdateRecipeRequest(
    string Title,
    string? DescriptionMarkdown,
    IReadOnlyList<RecipeIngredientRequest> Ingredients,
    int Servings = 1,
    string? MealType = null,
    string? SourceUrl = null,
    string? VideoUrl = null,
    string? VideoThumbnailUrl = null);

public sealed record RecipeIngredientResponse(
    Guid Id,
    int SortOrder,
    Guid IngredientId,
    string Name,
    decimal Quantity,
    string Unit,
    decimal NormalizedQuantity,
    string NormalizedUnit,
    string? Note);

public sealed record RecipeSummaryResponse(
    Guid Id,
    string Title,
    string Slug,
    Guid CreatedByUserId,
    int IngredientCount,
    int Servings,
    string? MealType,
    string? DescriptionMarkdown,
    string? FirstImageUrl,
    bool IsAiAssisted,
    DateTimeOffset CreatedAtUtc,
    DateTimeOffset UpdatedAtUtc);

public sealed record RecipeSummaryData(
    Guid Id,
    string Title,
    string Slug,
    Guid OwnerUserId,
    int IngredientCount,
    ServingCount Servings,
    MealSlot? MealType,
    string? DescriptionMarkdown,
    bool IsAiAssisted,
    DateTimeOffset CreatedAtUtc,
    DateTimeOffset UpdatedAtUtc)
{
    public static RecipeSummaryData FromRecipe(Recipe recipe) => new(
        recipe.Id,
        recipe.Title,
        recipe.Slug,
        recipe.OwnerUserId,
        recipe.Ingredients.Count,
        recipe.Servings,
        recipe.MealType,
        recipe.DescriptionMarkdown,
        recipe.IsAiAssisted,
        recipe.CreatedAtUtc,
        recipe.UpdatedAtUtc);
}

public sealed record RecipePlanningIngredientResponse(Guid Id, string Name);

public sealed record RecipePlanningSummaryResponse(
    Guid Id,
    string Title,
    int Servings,
    string? MealType,
    IReadOnlyList<RecipePlanningIngredientResponse> Ingredients,
    string? DescriptionMarkdown,
    string? FirstImageUrl);

public sealed record RecipePlanningSearchResponse(
    IReadOnlyList<RecipePlanningSummaryResponse> Recipes,
    int TotalCount);

public sealed record RecipeDetailResponse(
    Guid Id,
    string Title,
    string Slug,
    Guid CreatedByUserId,
    string? DescriptionMarkdown,
    int Servings,
    string? MealType,
    IReadOnlyList<RecipeIngredientResponse> Ingredients,
    IReadOnlyList<RecipeImageResponse> Images,
    bool CanEdit,
    bool IsAiAssisted,
    string? SourceUrl,
    string? VideoUrl,
    string? VideoThumbnailUrl,
    DateTimeOffset CreatedAtUtc,
    DateTimeOffset UpdatedAtUtc);

public sealed record PagedResponse<T>(
    IReadOnlyList<T> Items,
    int Page,
    int PageSize,
    int TotalCount);

public enum RecipeMatchType
{
    SourceUrl,
    ExactTitle,
    SimilarTitle
}

public sealed record RecipeExistenceMatch(
    Guid RecipeId,
    string Title,
    string? SourceUrl,
    RecipeMatchType MatchType);

public sealed record RecipeExistenceResult(
    bool Exists,
    IReadOnlyList<RecipeExistenceMatch> Matches);

public enum RecipeOperationStatus
{
    Success,
    NotFound,
    ValidationFailed
}

public sealed record RecipeOperationResult(
    RecipeOperationStatus Status,
    RecipeDetailResponse? Recipe = null,
    IReadOnlyDictionary<string, string[]>? ValidationErrors = null);
